stan/math/gpu/multiply.hpp

```diff
@@ -91,6 +91,23 @@ inline matrix_gpu multiply(const matrix_gpu& A, const matrix_gpu& B) {
   temp.sub_block(tempPad, 0, 0, 0, 0, temp.rows(), temp.cols());
   return temp;
 }
+
+/**
+ * Templated product operator for GPU matrices.
+ *
+ * Computes the matrix multiplication C[M, K] = A[M, N] x B[N, K]
+ *
+ * @param A A matrix or scalar
+ * @param B A matrix or scalar
+ * @return the product of the first and second arguments
+ *
+ * @throw <code>std::invalid_argument</code> if the
+ *   number of columns in A and rows in B do not match
+ */
+template <typename T1, typename T2>
+inline matrix_gpu operator*(const T1& A, const T2& B) {
```

Comment:
Why is this templated? Aren't they always matrix_gpus?

Reply:
For this particular operator, you can have the left or right side be a double. I was being lazy, but I think it's fine because it will only template when the return type is `matrix_gpu`?

If not then writing these signatures out is fine by me.

Reply:
I don't think that C++ is capable of using the return type in template deduction: https://stackoverflow.com/questions/8165659/why-cant-c-deduce-template-type-from-assignment

So I think this could get pulled in at weird times... not sure, might need to run some tests and make sure it doesn't mix poorly with normal matrix operations. Could you perhaps add some of those? I would say something that imports both gpu/multiply and rev/mat/multiply and tries operator overloading with a variety of types.

Reply:
Oh whoops totally my bad, I assumed so. I'll add some test types. Totally fine with adding the particular signatures here as well

Reply:
I think some simple tests of this will be good anyway! But yeah, they might reveal that we need to add the particular signatures (or more specific templated versions).

Reply:
Sorry for my delay on this, busy work week. I removed the templating and added particular signatures for each one.